File: 백준/Platinum/31914. 기초마법학/기초마법학.cc
#include<bits/stdc++.h>
using namespace std;
using ll = long long;

int main(){
    cin.tie(0); cout.tie(0); ios::sync_with_stdio(false);
    int n, k; cin >> n >> k;
    priority_queue<tuple<int, ll, int>> pq;
    vector<int> cnt(k+1, 0);
    vector<tuple<int, int, ll, int>> g;
    g.reserve(n);
    for(int i = 0; i < n; i++){
        int x, y, c; ll p; cin >> x >> y >> p >> c;
        g.push_back({x, y, p, c});
    }
    sort(g.begin(), g.end());
    ll ans, cur; ans = cur = 0;
    int maxh = 123456789;
    for(int i = 0; i < n; i++){
        auto [x, y, p, c] = g[i];
        pq.push({y, p, c});
        cur += p;
        if(!cnt[c]) cnt[0]++;
        cnt[c]++;
        if(i == n-1 || get<0>(g[i+1]) != x){
            while(pq.size()){
                auto [yy, pp, cc] = pq.top();
                if(cnt[0] != k && yy <= maxh) break;
                pq.pop();
                cur -= pp;
                cnt[cc]--;
                if(!cnt[cc]) cnt[0]--;
                maxh = min(maxh, yy-1);
            }
            ans = max(ans, cur);
        }
    }
    cout << ans;
    return 0;
}

File: 백준/Platinum/31914. 기초마법학/기초마법학.py
import sys
print = lambda *args, sep=" ", end="\n": sys.stdout.write(sep.join(map(str, args)) + end)
input = lambda: sys.stdin.readline().rstrip('\r\n')

import heapq

def main():
    n, k = map(int, input().split())
    cnt = [0]*(k+1)
    g = [tuple(map(int, input().split())) for _ in range(n)]
    g.sort()
    ans = cur = 0
    maxh = 123456789
    max_hq = []
    for i in range(n):
        x, y, p, c = g[i]
        heapq.heappush(max_hq, (-y, p, c))
        cur += p
        if not cnt[c]: cnt[0] += 1
        cnt[c] += 1
        if i == n-1 or g[i+1][0] != x:
            while max_hq:
                yy = -max_hq[0][0]
                if cnt[0] != k and yy <= maxh: break
                _, pp, cc = heapq.heappop(max_hq)
                cur -= pp
                cnt[cc] -= 1
                if not cnt[cc]: cnt[0] -= 1
                maxh = min(maxh, yy-1)
            ans = max(ans, cur)    
    print(ans)
    return 0
if __name__ == '__main__':
    sys.exit(main())